20220831-add-type-restrictions-to-json-syntax.md

```diff
@@ -0,0 +1,558 @@
+        "relation-6": { "directly_related_user_types": [{ "type": "user" }] }, // invalid, relation-6 does not allow direct relationships (no `this` in the relation definition)
+        "relation-7": { "directly_related_user_types": [] }, // valid, relation-7 does not allow direct relationships, so no elements are expected
+      }
+    }
+  }
+] }
+```
+
+#### Validating Tuple Writes
+
+We'll use the following example authorization model:
+
+```json
+{ "schema_version": "1.1",
+  "type_definitions": [
+  { "type": "user" },
+  { "type": "group",
+    "relations": {
+      "parent": { "this": {} },
+      "member": { "this": {} },
+    },
+    "metadata": {
+      "relations": {
+        "parent": { "directly_related_user_types": [{ "type": "group" }] },
+        "member": { "directly_related_user_types": [{ "type": "user" }, { "type": "group", "relation": "member" }, { "type": "user", "any": true }] }
+      }
+    }
+  },
+] }
+```
+
+On write, we need to validate that:
+
+1. If the user is an object:
+   1. The type of the user should be in the list of directly related user types on the relation of the type of the object
+   - `write(user=user:1, member, group:1)`; valid, the `user` type is in the list of directly related user types for the `member` relation of the `group` type
+   - `write(user=group:2, parent, group:1)`; valid, the `group` type is in the list of directly related user types for the `parent` relation of the `group` type
+   - `write(user=group:2, member, group:1)`; invalid, the `group` type is not in the list of directly related user types for the `member` relation of the `group` type
+   - `write(user=user:1, parent, group:1)`; invalid, the `user` type is not in the list of directly related user types for the `parent` relation of the `group` type
+
+2. If the user is a userset:
+   - `write(user=group:2#member, member, group:1)`; valid, the `(type=group, relation=member)` is in the list of directly related user types for the `member` relation of the `group` type
+   - `write(user=group:2#member, parent, group:1)`; invalid, the `(type=group, relation=member)` is not in the list of directly related user types for the `parent` relation of the `group` type
+   - `write(user=group:2#parent, member, group:1)`; invalid, the `(type=group, relation=parent)` is not in the list of directly related user types for the `member` relation of the `group` type
+   - `write(user=group:2#parent, parent, group:1)`; invalid, the `(type=group, relation=parent)` is not in the list of directly related user types for the `parent` relation of the `group` type
+
+3. If the user is `*`:
+   - `write(user=*, parent, group:1)`; invalid, the `parent` relation on the `group` type does not allow for any `*` relation
+   - `write(user=*, member, group:1)`; valid, the `member` relation on the `group` type allows for the `*` relation (and it will be considered all objects of type `user`)
```

Comment:
I thought that the `write(*, member, group:1)` syntax wouldn't be valid anymore and you'd have to `write(user:*, member, group:1)`

If this not the case then why do we have a `type` in  `{ "type": "user", "any": true }` ?

Reply:
Yes, I am a bit confused here too.

Reply:
* is valid, if any of the directly related user types has `any:true` in it.

You cannot write user:*, as that is an opject of type `user` and id `*`

Reply:
As discussed offline, should we consider restricting `*` so that in the future, we can allow restrict 
write(user=*, parent, group:1)
write(admin=*, parent, group:1) 
So that in the future, we can decide to restrict * to a particular type.  And in the very least, prevent confusion and mistake.

Reply:
We can decide to punt this to a later PR to keep the scope of RFC smaller